src/math/lp/nla_core.cpp

```diff
@@ -1560,9 +1568,15 @@ lbool core::check() {
     if (no_effect())
         m_monomial_bounds.propagate();
 
-    if (no_effect() && improve_bounds())
+    if (no_effect())  
+        m_monomial_bounds.improve_bounds();
+    if (lra.get_status() == lp::lp_status::INFEASIBLE) {
+        TRACE("nla_solver", tout << "return unknown because of the status = " << lra.get_status() <<"\n";);        
         return l_false;
-    
+    }
+    if (!integrality_holds()) {
```

Comment:
l_false should mean that there is a lemma, otherwise this can loop

Reply:
consider what the return modalities are: should there be a return modality that says to require LIA?
If LIA is invoked, does it ensure convergence? I guess it does because integrality is broken only if bounds are tightened. Though it is subtle with integer bounds.

Reply:
possibly disable sequential invocations of bounds propagation within the same backtracking scope.